docs/src/cli/deploy-a-program.md

```diff
@@ -0,0 +1,217 @@
+authority is specified during program deployment, the default keypair is used as
+the authority.  This is why redeploying a program in the steps above didn't
+require an authority to be explicitly specified.
+
+The authority can be specified during deployment:
+
+```bash
+solana program deploy --upgrade-authority <UPGRADE_AUTHORITY_SIGNER> <PROGRAM_FILEPATH>
+```
+
+Or after deployment and using the default keypair as the current authority:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --new-upgrade-authority <NEW_UPGRADE_AUTHORITY>
+```
+
+Or after deployment and specifying the current authority:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --upgrade-authority <UPGRADE_AUTHORITY_SIGNER> --new-upgrade-authority <NEW_UPGRADE_AUTHORITY>
+```
+
+### Immutable programs
+
+A program can be marked immutable, which prevents all further redeployments by
+specifying the `--final` flag during deployment:
+
+```bash
+solana program deploy <PROGRAM_FILEPATH> --final
+```
+
+Or anytime after:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --final
+```
+
+### Dumping a program to a file
+
+The deployed program may be dumped back to a local file:
+
+```bash
+solana program dump <ACCOUNT_ADDRESS> <OUTPUT_FILEPATH>
+```
+
+The dumped file will be in the same as what was deployed, so in the case of a
+shared object, the dumped file will be a fully functional shared object.  Note
+that the `dump` command dumps the entire data space, which means the output file
+will have trailing zeros after the shared object's data up to `max_len`.
+Sometimes it is useful to dump and compare a program to ensure it matches a
+known program binary.  The original program file can be zero-extended, hashed,
+and compared to the hash of the dumped file.
+
+```bash
+$ solana dump <ACCOUNT_ADDRESS> dump.so
+$ cp original.so extended.so
+$ truncate -r dump.so extended.so
+$ sha256sum extended.so dump.so
+```
+
+### Using an intermediary Buffer account
+
+Instead of deploying directly to the program account, the program can be written
+to an intermediary buffer account.  Doing so can be useful for things like
+multi-entity governed programs.
+
+```bash
+solana progrma write-buffer <PROGRAM_FILEPATH>
+```
+
+Buffer accounts support authorities like program accounts:
+
+```bash
+solana program set-buffer-authority <BUFFER_ADDRESS> --new-upgrade-authority <NEW_UPGRADE_AUTHORITY>
+```
+
+One exception is that buffer accounts cannot be marked immutable like program
+accounts can, so they don't support `--final`.
+
+The buffer account, once entirely written, can be passed to `deploy` to deploy
+the program:
+
+```bash
+solana program deploy --program-id <PROGRAM_ADDRESS> --buffer <BUFFER_ADDRESS>
+```
+
+Note, the buffer's authority must match the program's upgrade authority.
+
+Buffers also support `show` and `dump` just like programs do.
+
+### Non-redeployable programs
```

Comment:
Non-upgradeable ?

Reply:
Haven't used "upgrade" or "upgradeable" in the entire doc, didn't want to introduce that terminology here.

Reply:
Hmm, you do use the word upgradeable twice in this section below. I see your point, though (and added a suggestion for one of those uses).
`Non-redeployable` is a bear, though, and not entirely accurate, since any program can be redeployed to a different address.